### PR TITLE
8324873: [lworld] implementation of value classes construction

### PR DESCRIPTION
**Pull request by vicente-romero-oracle**
implementing value classes construction approach as per JEP-401. There is also a bit of an off-road incursion into jdeps removing references to Q-descriptors, probably should have gone in another patch but the changes are kind of small anyways

---------
### Progress
- [x] Change must not contain extraneous whitespace

### Issue
 * [JDK-8324873](https://bugs.openjdk.org/browse/JDK-8324873): [lworld] implementation of value classes construction (**Bug** - P4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/valhalla.git pull/990/head:pull/990` \
`$ git checkout pull/990`

Update a local copy of the PR: \
`$ git checkout pull/990` \
`$ git pull https://git.openjdk.org/valhalla.git pull/990/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 990`

View PR using the GUI difftool: \
`$ git pr show -t 990`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/valhalla/pull/990.diff">https://git.openjdk.org/valhalla/pull/990.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/valhalla/pull/990#issuecomment-1923977544)

### CODE REVIEW

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Check.java [hunk 1]
```diff
@@ -4165,7 +4168,7 @@ public void visitApply(JCMethodInvocation apply) {
                 }
 
                 // If super()/this() isn't first, require "statements before super()" feature
-                if (!firstStatement)
+                if (!firstStatement && !classSym.isValueClass()) // let's wave value classes for now
```
Comment (archiecobbs) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Check.java:4171]: wave → waive

File: src/java.base/share/classes/java/lang/reflect/AccessFlag.java [hunk 2]
```diff
@@ -527,7 +532,7 @@ public static Set<AccessFlag> maskToAccessFlags(int mask, Location location) {
         Set<AccessFlag> result = java.util.EnumSet.noneOf(AccessFlag.class);
         for (var accessFlag : LocationToFlags.locationToFlags.get(location)) {
             int accessMask = accessFlag.mask();
-            if ((mask & accessMask) != 0) {
+            if ((mask &  accessMask) != 0) {
```
Comment (archiecobbs) [inline src/java.base/share/classes/java/lang/reflect/AccessFlag.java:535]: Extra whitespace glitch here.

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Check.java (hunk 1 above)
Comment (vicente-romero-oracle) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Check.java:4171]: oops

File: src/java.base/share/classes/java/lang/reflect/AccessFlag.java [hunk 3]
```diff
@@ -360,14 +360,19 @@ public Set<Location> apply(ClassFileFormatVersion cffv) {
...
                public Set<Location> apply(ClassFileFormatVersion cffv) {
-                   return (cffv.compareTo(ClassFileFormatVersion.RELEASE_2)  >= 0 &&
-                           cffv.compareTo(ClassFileFormatVersion.RELEASE_16) <= 0) ?
-                       Location.SET_METHOD:
-                       Location.EMPTY_SET;}
+                   if (ValhallaFeatures.isEnabled() && cffv.compareTo(ClassFileFormatVersion.RELEASE_22) >= 0) {
+                       return Location.SET_FIELD;
+                   } else {
+                       return (cffv.compareTo(ClassFileFormatVersion.RELEASE_2) >= 0 &&
+                               cffv.compareTo(ClassFileFormatVersion.RELEASE_16) <= 0) ?
```
Comment (liach) [inline src/java.base/share/classes/java/lang/reflect/AccessFlag.java:371]: We can probably change this ternary operator to be part of the if-else chain to make it clearer.

**Pull request closed without merging**